core/src/main/java/org/elasticsearch/index/seqno/LocalCheckpointService.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.index.seqno;
+
+import org.apache.lucene.util.FixedBitSet;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.shard.AbstractIndexShardComponent;
+import org.elasticsearch.index.shard.ShardId;
+
+/**
+ * This class generates sequences numbers and keeps track of the so called local checkpoint - the highest number for which
+ * all previous seqNo have been processed (including)
+ */
+public class LocalCheckpointService extends AbstractIndexShardComponent {
+
+    /** sets the maximum spread between lowest and highest seq no in flight */
+    public static String SETTINGS_INDEX_LAG_THRESHOLD = "index.seq_no.index_lag.threshold";
+
+    /**
+     * how long should an incoming indexing request which violates {@link #SETTINGS_INDEX_LAG_THRESHOLD } should wait
+     * before being rejected
+     */
+    public static String SETTINGS_INDEX_LAG_MAX_WAIT = "index.seq_no.index_lag.max_wait";
+
+    /** default value for {@link #SETTINGS_INDEX_LAG_THRESHOLD} */
+    final static int DEFAULT_INDEX_LAG_THRESHOLD = 1024;
+
+    /** default value for {@link #SETTINGS_INDEX_LAG_MAX_WAIT} */
+    final static TimeValue DEFAULT_INDEX_LAG_MAX_WAIT = TimeValue.timeValueSeconds(30);
+
+    /** protects changes to all internal state and signals changes in {@link #checkpoint} */
+    final Object mutex = new Object();
+
+    /** each bits maps to a seqNo in round robin fashion. a set bit means the seqNo has been processed */
```

Comment:
I'm not sure if "round robin" is the correct term here. Maybe just say each sequence number maps to a bit mod the lag?